zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java

```diff
@@ -341,41 +370,141 @@ IndexTemplates ensureIndexTemplates() {
       IndexTemplates templates = new VersionSpecificTemplates(this).get();
       HttpCall.Factory http = http();
-      ensureIndexTemplate(http, buildUrl(http, templates, SPAN), templates.span());
-      ensureIndexTemplate(http, buildUrl(http, templates, DEPENDENCY), templates.dependency());
-      ensureIndexTemplate(http, buildUrl(http, templates, AUTOCOMPLETE), templates.autocomplete());
+      ensureIndexTemplate(http, buildUrl(templates, SPAN), templates.span());
+      ensureIndexTemplate(http, buildUrl(templates, DEPENDENCY), templates.dependency());
+      ensureIndexTemplate(http, buildUrl(templates, AUTOCOMPLETE), templates.autocomplete());
       return templates;
     } catch (IOException e) {
       throw Platform.get().uncheckedIOException(e);
     }
   }
 
-  HttpUrl buildUrl(HttpCall.Factory http, IndexTemplates templates, String type) {
-    HttpUrl.Builder builder = http.baseUrl.newBuilder("_template");
+  String buildUrl(IndexTemplates templates, String type) {
     String indexPrefix = indexNameFormatter().index() + templates.indexTypeDelimiter();
-    return builder.addPathSegment(indexPrefix + type + "_template").build();
+    return "/_template/" + indexPrefix + type + "_template";
   }
 
-  @Memoized // hosts resolution might imply a network call, and we might make a new okhttp instance
-  public HttpCall.Factory http() {
+  @Memoized // a new client factory means new connections
+  ClientFactory clientFactory() {
+    ClientFactoryBuilder builder = new ClientFactoryBuilder()
+      // TODO(anuraaga): Remove after https://github.com/line/armeria/pull/1899
+      .workerGroup(EventLoopGroups.newEventLoopGroup(
+        Flags.numCommonWorkers(), "armeria-common-worker", true), true)
+      .useHttp2Preface(false);
+    clientFactoryCustomizer().accept(builder);
+    return builder.build();
+  }
+
+  @Memoized // hosts resolution might imply a network call, and we might make a new client instance
+  public HttpClient httpClient() {
     List<String> hosts = hostsSupplier().get();
     if (hosts.isEmpty()) throw new IllegalArgumentException("no hosts configured");
-    OkHttpClient ok =
-        hosts.size() == 1
-            ? client()
-            : client()
-                .newBuilder()
-                .dns(PseudoAddressRecordSet.create(hosts, client().dns()))
-                .build();
-    ok.dispatcher().setMaxRequests(maxRequests());
-    ok.dispatcher().setMaxRequestsPerHost(maxRequests());
-    return new HttpCall.Factory(ok, HttpUrl.parse(hosts.get(0)));
+
+    List<URL> urls = hosts.stream()
+      .map(host -> {
+        try {
+          return new URL(host);
+        } catch (MalformedURLException e) {
+          throw new IllegalArgumentException("Invalid host: " + host, e);
+        }
+      })
+      .collect(Collectors.toList());
+
+    final EndpointGroup endpointGroup;
+    if (urls.size() == 1) {
+      URL url = urls.get(0);
+      if (isIpAddress(url.getHost())) {
+        endpointGroup = null;
```

Comment:
logic like this needs a comment. my guess is that a hostname could be potentially more than one IPs and that's why. Just a few word trailing comment in the main branches here would be handy.